File: app/src/main/java/com/ishmah/musichub/db/UserProfileDao.java
package com.ishmah.musichub.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.os.Handler;
import android.os.Looper;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class UserProfileDao {

    public interface IntCallback {
        void onResult(int value);
    }

    private final android.database.sqlite.SQLiteDatabase db;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    public UserProfileDao(Context context) {
        db = new DatabaseHelper(context).getWritableDatabase();
    }

    public void saveProfile(String username, String bio, String photoUri,
                            String theme, boolean notifEnabled, Runnable onDone) {
        executor.execute(() -> {
            Cursor cursor = db.query(DatabaseHelper.TABLE_USER_PROFILE,
                    null, null, null, null, null, null);
            boolean exists = cursor.getCount() > 0;
            cursor.close();

            ContentValues values = new ContentValues();
            values.put(DatabaseHelper.COL_USERNAME, username);
            values.put(DatabaseHelper.COL_BIO, bio);
            values.put(DatabaseHelper.COL_PHOTO_URI, photoUri);
            values.put(DatabaseHelper.COL_THEME, theme);
            values.put(DatabaseHelper.COL_NOTIF_ENABLED, notifEnabled ? 1 : 0);

            if (exists) {
                db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null);
            } else {
                db.insert(DatabaseHelper.TABLE_USER_PROFILE, null, values);
            }
            if (onDone != null) onDone.run();
        });
    }

    public Map<String, String> getProfile() {
        Cursor cursor = db.query(DatabaseHelper.TABLE_USER_PROFILE,
                null, null, null, null, null, null);
        if (!cursor.moveToFirst()) {
            cursor.close();
            return null;
        }
        Map<String, String> profile = new HashMap<>();
        profile.put("username", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_USERNAME)));
        profile.put("bio", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_BIO)));
        profile.put("photo_uri", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_PHOTO_URI)));
        profile.put("theme", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_THEME)));
        profile.put("notif_enabled", cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_NOTIF_ENABLED)));
        cursor.close();
        return profile;
    }

    public void updateUsername(String username, Runnable onDone) {
        executor.execute(() -> {
            ContentValues values = new ContentValues();
            values.put(DatabaseHelper.COL_USERNAME, username);
            db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null);
            if (onDone != null) onDone.run();
        });
    }

    public void updatePhoto(String photoUri, Runnable onDone) {
        executor.execute(() -> {
            ContentValues values = new ContentValues();
            values.put(DatabaseHelper.COL_PHOTO_URI, photoUri);
            db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null);
            if (onDone != null) onDone.run();
        });
    }

    public void updateTheme(String theme, Runnable onDone) {
        executor.execute(() -> {
            ContentValues values = new ContentValues();
            values.put(DatabaseHelper.COL_THEME, theme);
            db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null);
            if (onDone != null) onDone.run();
        });
    }

    public void addListeningTime(int seconds) {
        if (seconds <= 0) return;
        executor.execute(() -> {
            Cursor cursor = db.query(DatabaseHelper.TABLE_USER_PROFILE,
                    new String[]{DatabaseHelper.COL_LISTENING_SECONDS},
                    null, null, null, null, null);
            boolean hasRow = cursor.moveToFirst();
            int current = hasRow ? cursor.getInt(0) : 0;
            cursor.close();

            ContentValues values = new ContentValues();
            values.put(DatabaseHelper.COL_LISTENING_SECONDS, current + seconds);

            if (hasRow) {
                db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null);
            } else {
                // No profile row yet — insert one so the time isn't lost
                db.insert(DatabaseHelper.TABLE_USER_PROFILE, null, values);
            }
        });
    }

    public void getTotalListeningSeconds(IntCallback callback) {
        executor.execute(() -> {
            Cursor cursor = db.query(DatabaseHelper.TABLE_USER_PROFILE,
                    new String[]{DatabaseHelper.COL_LISTENING_SECONDS},
                    null, null, null, null, null);
            int total = 0;
            if (cursor.moveToFirst()) {
                total = cursor.getInt(0);
            }
            cursor.close();
            int finalTotal = total;
            new Handler(Looper.getMainLooper()).post(() -> callback.onResult(finalTotal));
        });
    }
}

File: app/src/main/java/com/ishmah/musichub/db/UserProfileDao.kt
package com.ishmah.musichub.db

import android.content.ContentValues
import android.content.Context
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class UserProfileDao(context: Context) {

    private val db = DatabaseHelper(context).writableDatabase
    private val executor: Executor = Executors.newSingleThreadExecutor()

    // Simpan / update profile
    fun saveProfile(
        username: String,
        bio: String,
        photoUri: String,
        theme: String,
        notifEnabled: Boolean,
        onDone: (() -> Unit)? = null
    ) {
        executor.execute {
            // Cek apakah sudah ada data profile
            val cursor = db.query(
                DatabaseHelper.TABLE_USER_PROFILE,
                null, null, null, null, null, null
            )
            val exists = cursor.count > 0
            cursor.close()

            val values = ContentValues().apply {
                put(DatabaseHelper.COL_USERNAME, username)
                put(DatabaseHelper.COL_BIO, bio)
                put(DatabaseHelper.COL_PHOTO_URI, photoUri)
                put(DatabaseHelper.COL_THEME, theme)
                put(DatabaseHelper.COL_NOTIF_ENABLED, if (notifEnabled) 1 else 0)
            }

            if (exists) {
                // Update kalau sudah ada
                db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null)
            } else {
                // Insert kalau belum ada
                db.insert(DatabaseHelper.TABLE_USER_PROFILE, null, values)
            }
            onDone?.invoke()
        }
    }

    // Ambil profile
    fun getProfile(): Map<String, String>? {
        val cursor = db.query(
            DatabaseHelper.TABLE_USER_PROFILE,
            null, null, null, null, null, null
        )
        if (!cursor.moveToFirst()) {
            cursor.close()
            return null
        }
        val profile = mapOf(
            "username" to (cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_USERNAME)) ?: ""),
            "bio" to (cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_BIO)) ?: ""),
            "photo_uri" to (cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_PHOTO_URI)) ?: ""),
            "theme" to (cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_THEME)) ?: "dark"),
            "notif_enabled" to cursor.getString(cursor.getColumnIndexOrThrow(DatabaseHelper.COL_NOTIF_ENABLED))
        )
        cursor.close()
        return profile
    }

    // Update username saja
    fun updateUsername(username: String, onDone: (() -> Unit)? = null) {
        executor.execute {
            val values = ContentValues().apply {
                put(DatabaseHelper.COL_USERNAME, username)
            }
            db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null)
            onDone?.invoke()
        }
    }

    // Update foto profil saja
    fun updatePhoto(photoUri: String, onDone: (() -> Unit)? = null) {
        executor.execute {
            val values = ContentValues().apply {
                put(DatabaseHelper.COL_PHOTO_URI, photoUri)
            }
            db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null)
            onDone?.invoke()
        }
    }

    // Update theme saja
    fun updateTheme(theme: String, onDone: (() -> Unit)? = null) {
        executor.execute {
            val values = ContentValues().apply {
                put(DatabaseHelper.COL_THEME, theme)
            }
            db.update(DatabaseHelper.TABLE_USER_PROFILE, values, null, null)
            onDone?.invoke()
        }
    }
}